Fix warnings issued by Intel OneAPI 2021.4.0

KineticTheory opened the pull request:
### Background

* We are adding support for Intel OneAPI compilers for Crossroads.  This change-set eliminates build warnings for `Release` builds. 
* These changes seem to also eliminate all build warnings for `Debug` flavor.

### Purpose of Pull Request

* [Fixes re-git issue #56](https://re-git.lanl.gov/draco/devops/-/issues/56)

### Description of changes

* Always build with `-fp-model=precise`
* Provide copy constructors and copy assignment operators for classes that have explicit destructors.
* Remove variables that are defined and set, but not read.
* Add `pragmas` to quiet warnings for vendor code.

### Status

* Reference:
  * [Pre-Merge Code Review](https://github.com/lanl/Draco/wiki/Style-Guide)
  * [CDash Status](https://rtt.lanl.gov/cdash3/index.php?project=Draco&filtercount=1&showfilters=1&field1=buildname&compare1=63&value1=-pr)
* Checks
  * [x] Travis/Appveyor CI checks pass
  * [x] Code coverage does not decrease
  * [x] [Valgrind test passes](https://rtt.lanl.gov/cdash3/index.php?project=Draco)
  * [x] [Toss3 checks pass](https://rtt.lanl.gov/cdash3/index.php?project=Draco)
  * [x] [Trinitite checks pass](https://rtt.lanl.gov/cdash3/index.php?project=Draco)
  * [ ] Code reviewed/approved, sufficient DbC checks, testing, documentation

KineticTheory committed:
Addressing warnings reported by OneAPI 2021.4.0.

* Always build with `-fp-model=precise`
* Provide copy constructors and copy assignment operators for classes that have explicit
  destructors.
+ Remove variables that are defined and set, but not read.
+ Add pragmas to quiet warnings for vendor code.

KineticTheory committed:
Remove a debug print

certik commented:
There is still one warning, treated as error:
```
[ 25%] Building CXX object src/linear/CMakeFiles/Lib_linear.dir/qrdcmp_pt.cc.o
/home/travis/Draco/src/mesh_element/Element_Definition.hh:262:3: error: use '= default' to define a trivial copy constructor [modernize-use-equals-default,-warnings-as-errors]
  Element_Definition(Element_Definition const &rhs)
  ^
2810 warnings generated.
Suppressed 2809 warnings (2809 in non-user code).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
1 warning treated as error
make[2]: *** [src/mesh_element/CMakeFiles/Lib_mesh_element.dir/build.make:82: src/mesh_element/CMakeFiles/Lib_mesh_element.dir/Element_Definition.cc.o] Error 1
```
Otherwise it looks good.

KineticTheory committed:
Fix a clang-tidy warning.

KineticTheory commented:
> Otherwise it looks good.

Should be fixed by c156e7f.

KineticTheory commented on src/kde/quick_index.cc:
```diff
@@ -456,7 +455,7 @@ quick_index::window_coarse_index_list(const std::array<double, 3> &window_min,
     // Only one bound of the window should every overshoot zero
     Check(!(window_min[1] < 0.0 && window_max[1] > 2.0 * rtt_units::PI));
-    size_t overlap_nbins = 1;
+    // size_t overlap_nbins = 1;
```
@clevelam Can you confirm that this variable is not used?

KineticTheory replied:
> 
> 
> > Otherwise it looks good.
> 
> Should be fixed by [c156e7f](https://github.com/lanl/Draco/commit/c156e7f6931f70a0f164270c84af24f3a73f84a1).

Early exit got me.  Let me run the whole set of `clang-tidy` before submitting another update.

KineticTheory reviewed (commented):
Leaving an question for Matt.

KineticTheory committed:
Fix clang-tidy warnings.

+ Also attempt use use "rule of 5" for cases where one or more copy/move ctor or assignment operator
  are defined.
+ Use `= default` whenever possible for copy/move ctor or assignment operators.
